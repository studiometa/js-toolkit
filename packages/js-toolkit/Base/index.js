import { getComponentElements } from './utils.js';
import ChildrenManager from './managers/ChildrenManager.js';
import RefsManager from './managers/RefsManager.js';
import ServicesManager from './managers/ServicesManager.js';
import EventsManager from './managers/EventsManager.js';
import OptionsManager from './managers/OptionsManager.js';

let id = 0;

// eslint-disable-next-line no-undef
const isDev = typeof __DEV__ !== 'undefined' && __DEV__;

/**
 * No operation function.
 * @returns {void}
 */
function noop() {}

/**
 * Test if the managers' instances implement the default manager.
 *
 * @throws
 * @param   {Base} instance The instance to test.
 * @returns {void}
 */
function createAndTestManagers(instance) {
  [
    {
      prop: '__options',
      constructorName: 'OptionsManager',
      constructor: OptionsManager,
    },
    {
      prop: '__services',
      constructorName: 'ServicesManager',
      constructor: ServicesManager,
    },
    {
      prop: '__events',
      constructorName: 'EventsManager',
      constructor: EventsManager,
    },
    {
      prop: '__refs',
      constructorName: 'RefsManager',
      constructor: RefsManager,
    },
    {
      prop: '__children',
      constructorName: 'ChildrenManager',
      constructor: ChildrenManager,
    },
  ].forEach(({ prop, constructorName, constructor }) => {
    instance[prop] = new instance.__managers[constructorName](instance);
    if (!(instance[prop] instanceof constructor)) {
      throw new Error(
        `The \`$managers.${constructorName}\` must extend the \`${constructorName}\` class.`
      );
    }
  });
}

/**
 * @typedef {typeof Base} BaseConstructor
 * @typedef {(Base) => Promise<BaseConstructor | { default: BaseConstructor }>} BaseAsyncConstructor
 * @typedef {OptionsManager & { [name:string]: any }} BaseOptions
 * @typedef {{ [name:string]: HTMLElement | HTMLElement[] }} BaseRefs
 * @typedef {{ [nameOrSelector:string]: Base[] | Promise<Base>[] }} BaseChildren
 * @typedef {{ [nameOrSelector:string]: BaseConstructor | BaseAsyncConstructor }} BaseConfigComponents
 * @typedef {import('./managers/OptionsManager').OptionsSchema} BaseConfigOptions
 * @typedef {import('./managers/ServicesManager.js').ServiceName} ServiceName
 */

/**
 * @typedef {Object} BaseConfig
 * @property {string} name
 * @property {boolean} [debug]
 * @property {boolean} [log]
 * @property {string[]} [refs]
 * @property {string[]} [emits]
 * @property {BaseConfigComponents} [components]
 * @property {BaseConfigOptions} [options]
 */

/**
 * @typedef {Object} Managers
 * @property {typeof ChildrenManager} ChildrenManager
 * @property {typeof EventsManager} EventsManager
 * @property {typeof OptionsManager} OptionsManager
 * @property {typeof RefsManager} RefsManager
 * @property {typeof ServicesManager} ServicesManager
 */

/**
 * Base class.
 */
export default class Base extends EventTarget {
  /**
   * This is a Base instance.
   * @type {boolean}
   * @readonly
   */
  static $isBase = true;

  /**
   * The instance parent.
   * @type {Base}
   */
  $parent = null;

  /**
   * The instance id.
   * @type {string}
   */
  $id;

  /**
   * The root element.
   * @type {HTMLElement & { __base__?: WeakMap<BaseConstructor, Base | 'terminated'> }}
   */
  $el;

  /**
   * The state of the component.
   * @type {boolean}
   */
  $isMounted = false;

  /**
   * Store the event handlers.
   *
   * @private
   * @type {Map<string, Set<EventListenerOrEventListenerObject>>}
   */
  __eventHandlers = new Map();

  /**
   * Get the root instance of the app.
   * @returns {Base}
   */
  get $root() {
    if (!this.$parent) {
      return this;
    }

    let parent = this.$parent;
    let root = this.$parent;

    while (parent) {
      if (!parent.$parent) {
        root = parent;
      }

      parent = parent.$parent;
    }

    return root;
  }

  /**
   * Merge configuration with the parents' configurations.
   *
   * @returns {BaseConfig}
   * @private
   */
  get __config() {
    let proto = Object.getPrototypeOf(this);
    let { config } = proto.constructor;

    while (proto.constructor.config && proto.constructor.$isBase) {
      config = { ...proto.constructor.config, ...config };

      if (proto.constructor.config.options) {
        config.options = { ...proto.constructor.config.options, ...config.options };
      }

      if (proto.constructor.config.emits && config.emits) {
        config.emits = [...proto.constructor.config.emits, ...config.emits];
      }

      proto = Object.getPrototypeOf(proto);
    }

    config.options = config.options ?? {};
    config.refs = config.refs ?? [];
    config.components = config.components ?? {};

    return config;
  }

  /**
   * @type {BaseConfig}
   */
  static config = {
    name: 'Base',
    emits: [
      // hook events
      'mounted',
      'updated',
      'destroyed',
      'terminated',
      // default services' events
      'ticked',
      'scrolled',
      'resized',
      'moved',
      'loaded',
      'keyed',
    ],
  };

  /**
   * @type {ServicesManager}
   * @private
   */
  __services;

  /**
   * @type {ServicesManager}
   */
  get $services() {
    return this.__services;
  }

  /**
   * @type {RefsManager}
   * @private
   */
  __refs;

  /**
   * @returns {RefsManager}
   */
  get $refs() {
    return this.__refs;
  }

  /**
   * @type {BaseOptions}
   * @private
   */
  __options;

  /**
   * @returns {BaseOptions}
   */
  get $options() {
    return this.__options;
  }

  /**
   * @type {ChildrenManager}
   * @private
   */
  __children;

  /**
   * @returns {ChildrenManager}
   */
  get $children() {
    return this.__children;
  }

  /**
   * @type {EventsManager}
   * @private
   */
  __events;

  /**
   * Small helper to log stuff.
   *
   * @returns {(...args: any) => void}
   */
  get $log() {
    return this.__options.log ? window.console.log.bind(window, `[${this.__config.name}]`) : noop;
  }

  /**
   * Small helper to debug information.
   *
   * @private
   * @returns {(...args:any) => void}
   */
  get __debug() {
    return isDev && this.__options.debug
      ? window.console.log.bind(window, `[debug] [${this.$id}]`)
      : noop;
  }

  /**
   * Get manager constructors.
   *
   * @returns {Managers}
   */
  get __managers() {
    return {
      ChildrenManager,
      EventsManager,
      OptionsManager,
      RefsManager,
      ServicesManager,
    };
  }

  /**
   * Call an instance method and emit corresponding events.
   *
   * @private
   * @param {string} method
   * @param {any[]} args
   * @returns {void}
   */
  __callMethod(method, ...args) {
    if (isDev) {
      this.__debug('callMethod', method, ...args);
    }

    this.$emit(method, ...args);

    // We always emit an event, but we do not call the method if it does not exist
    if (typeof this[method] !== 'function') {
      return;
    }

    this[method].call(this, ...args);
    if (isDev) {
      this.__debug(method, this, ...args);
    }
  }

  /**
   * Test if the given event has been bound to the instance.
   *
   * @private
   * @param  {string} event The event's name.
   * @returns {boolean}      Wether the given event has been bound or not.
   */
  __hasEvent(event) {
    const eventHandlers = this.__eventHandlers.get(event);
    return eventHandlers && eventHandlers.size > 0;
  }

  /**
   * Class constructor where all the magic takes place.
   *
   * @param {HTMLElement} element The component's root element dd.
   */
  constructor(element) {
    super();

    if (!element) {
      throw new Error('The root element must be defined.');
    }

    const { __config } = this;

    if (__config.name === 'Base') {
      throw new Error('The `config.name` property is required.');
    }

    this.$id = `${__config.name}-${id}`;
    id += 1;

    this.$el = element;

    if (!this.$el.__base__) {
      this.$el.__base__ = new WeakMap();
    }

    this.$el.__base__.set(this.__ctor, this);

    createAndTestManagers(this);

    if (isDev) {
      this.__debug('constructor', this);
    }
  }

  /**
   * Trigger the `mounted` callback.
   * @returns {this}
   */
  $mount() {
    if (this.$isMounted) {
      return this;
    }

    this.$isMounted = true;
    if (isDev) {
      this.__debug('$mount');
    }
    this.$children.registerAll();
    this.$refs.registerAll();
    this.__events.bindRootElement();
    this.$services.enableAll();
    this.$children.mountAll();

    this.__callMethod('mounted');

    return this;
  }

  /**
   * Update the instance children.
   * @returns {this}
   */
  $update() {
    if (isDev) {
      this.__debug('$update');
    }

    // Undo
    this.$refs.unregisterAll();
    this.$services.disableAll();

    // Redo
    this.$children.registerAll();
    this.$refs.registerAll();
    this.$services.enableAll();

    // Update
    this.$children.updateAll();

    this.__callMethod('updated');
    return this;
  }

  /**
   * Trigger the `destroyed` callback.
   * @returns {this}
   */
  $destroy() {
    if (!this.$isMounted) {
      return this;
    }

    this.$isMounted = false;
    if (isDev) {
      this.__debug('$destroy');
    }
    this.__events.unbindRootElement();
    this.$refs.unregisterAll();
    this.$services.disableAll();
    this.$children.destroyAll();
    this.__callMethod('destroyed');

    return this;
  }

  /**
   * Terminate a child instance when it is not needed anymore.
   * @returns {void}
   */
  $terminate() {
    if (isDev) {
      this.__debug('$terminate');
    }

    // First, destroy the component.
    this.$destroy();

    // Execute the `terminated` hook if it exists
    this.__callMethod('terminated');

    // Delete instance
    this.$el.__base__.set(this.__ctor, 'terminated');
  }

  /**
   * Test if the given event is configured and warn otherwise.
   *
   * @private
   * @param   {string} event The event name.
   * @returns {void}
   */
  __testEventIsDefined(event) {
    if (!(Array.isArray(this.__config.emits) && this.__config.emits.includes(event))) {
      console.warn(
        `[${this.__config.name}]`,
        `The "${event}" event is missing from the configuration.`
      );
    }
  }

  /**
   * Add an emitted event.
   *
   * @private
   * @param   {string} event The event name.
   * @returns {void}
   */
  __addEmits(event) {
    const ctor = this.__ctor;
    if (Array.isArray(ctor.config.emits)) {
      ctor.config.emits.push(event);
    } else {
      ctor.config.emits = [event];
    }
  }

  /**
   * Remove an emitted event.
   *
   * @private
   * @param   {string} event The event name.
   * @returns {void}
   */
  __removeEmits(event) {
    const ctor = this.__ctor;
    const index = ctor.config.emits.indexOf(event);
    ctor.config.emits.splice(index, 1);
  }

  /**
   * Get the instance constructor.
   *
   * @private
   * @returns {BaseConstructor}
   */
  get __ctor() {
    return /** @type {BaseConstructor} */ (this.constructor);
  }

  /**
   * Bind a listener function to an event.
   *
   * @param  {string} event
   *   Name of the event.
   * @param  {EventListenerOrEventListenerObject} listener
   *   Function to be called.
   * @param {boolean|AddEventListenerOptions} [options]
   *   Options for the `removeEventListener` method.
   * @returns {() => void}
   *   A function to unbind the listener.
   */
  $on(event, listener, options) {
    if (isDev) {
      this.__debug('$on', event, listener, options);
    }

    if (isDev) {
      this.__testEventIsDefined(event);
    }

    let set = this.__eventHandlers.get(event);

    if (!set) {
      set = new Set();
      this.__eventHandlers.set(event, set);
    }

    set.add(listener);
    this.addEventListener(event, listener, options);

    return () => {
      this.$off(event, listener, options);
    };
  }

  /**
   * Unbind a listener function from an event.
   *
   * @param {string} event
   *   Name of the event.
   * @param {EventListenerOrEventListenerObject} listener
   *   Function to be removed.
   * @param {boolean|EventListenerOptions} [options]
   *   Options for the `removeEventListener` method.
   * @returns {void}
   */
  $off(event, listener, options) {
    if (isDev) {
      this.__debug('$off', event, listener);
    }

    if (isDev) {
      this.__testEventIsDefined(event);
    }

    this.__eventHandlers.get(event).delete(listener);
    this.removeEventListener(event, listener, options);
  }

  /**
   * Emits an event.
   *
   * @param  {string} event
   *   Name of the event.
   * @param  {any[]}        args  The arguments to apply to the functions bound to this event.
   * @returns {void}
   */
  $emit(event, ...args) {
    if (isDev) {
      this.__debug('$emit', event, args);
    }

    if (isDev) {
      this.__testEventIsDefined(event);
    }

    this.dispatchEvent(new CustomEvent(event, { detail: args }));
  }

  /**
   * Factory method to generate multiple instance of the class.
   *
   * @param  {string}      nameOrSelector The selector on which to mount each instance.
   * @returns {Array<Base>}                A list of the created instance.
   */
  static $factory(nameOrSelector) {
    if (!nameOrSelector) {
      throw new Error(
        'The $factory method requires a component’s name or selector to be specified.'
      );
    }

    return getComponentElements(nameOrSelector).map((el) => new this(el).$mount());
  }
}
