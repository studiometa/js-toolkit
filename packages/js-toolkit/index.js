export { default as Base } from './Base/index.js';
export * from './decorators/index.js';
export * from './helpers/index.js';
export * from './services/index.js';
