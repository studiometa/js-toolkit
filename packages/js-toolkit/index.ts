export { Base } from './Base/index.js';
export * from './decorators/index.js';
export * from './helpers/index.js';
export * from './services/index.js';

export type {
  BaseTypeParameter,
  BaseConstructor,
  BaseAsyncConstructor,
  BaseConfig,
} from './Base/index.js';

export type { BaseInterface, BaseDecorator } from './Base/types.js';
