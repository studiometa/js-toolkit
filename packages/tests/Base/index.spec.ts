import { describe, it, expect, spyOn, mock, beforeEach, afterEach } from 'bun:test';
import { Base, BaseConfig, BaseProps, getInstanceFromElement } from '@studiometa/js-toolkit';
import { ChildrenManager, OptionsManager, RefsManager } from '#private/Base/managers/index.js';
import { h, useFakeTimers, useRealTimers, advanceTimersByTimeAsync } from '#test-utils';

beforeEach(() => useFakeTimers());
afterEach(() => useRealTimers());

async function getContext() {
  class Foo<T extends BaseProps = BaseProps> extends Base<T> {
    static config = {
      name: 'Foo',
    };
  }
  const element = h('div');
  const foo = new Foo(element).$mount();
  await advanceTimersByTimeAsync(1);

  return { Foo, element, foo };
}

describe('The abstract Base class', () => {
  it('must be extended', () => {
    expect(() => {
      new Base(document.createElement('div'));
    }).toThrow();
  });

  it('should throw an error when extended without proper configuration', () => {
    expect(() => {
      // @ts-ignore
      class Foo extends Base {
        static config = {};
      }
      new Foo(document.createElement('div'));
    }).toThrow('The `config.name` property is required.');
  });

  it('should throw an error if instantiated without a root element.', () => {
    expect(() => {
      class Foo extends Base {
        static config = {
          name: 'Foo',
        };
      }
      // @ts-ignore
      new Foo();
    }).toThrow('The root element must be defined.');
  });
});

describe('A Base instance', () => {
  it('should have an `$id` property', async () => {
    const { foo } = await getContext();
    expect(foo.$id).toBeDefined();
  });

  it('should have an `$isMounted` property', async () => {
    const { foo } = await getContext();
    expect(foo.$isMounted).toBe(true);
  });

  it('should have a `$refs` property', async () => {
    const { foo } = await getContext();
    expect(foo.$refs).toBeInstanceOf(RefsManager);
  });

  it('should have a `$children` property', async () => {
    const { foo } = await getContext();
    expect(foo.$children).toBeInstanceOf(ChildrenManager);
  });

  it('should have an `$options` property', async () => {
    const { foo } = await getContext();
    expect(foo.$options).toBeInstanceOf(OptionsManager);
    expect(foo.$options.name).toBe('Foo');
  });

  it('should be able to set any `$options` property', async () => {
    const { foo } = await getContext();
    foo.$options.log = true;
    expect(foo.$options.log).toBe(true);
  });

  it('should have an `$el` property', async () => {
    const { foo, element } = await getContext();
    expect(foo.$el).toBe(element);
  });

  it('should have a `__base__` property', async () => {
    const { foo, Foo } = await getContext();
    // @ts-ignore
    expect(foo.$el.__base__).toBeInstanceOf(WeakMap);
    expect(foo.$el.__base__.get(Foo)).toBe(foo);
  });

  it('should inherit from parent config', () => {
    class A extends Base {
      static config = {
        name: 'A',
        log: true,
      };
    }

    class B extends A {
      static config = {
        name: 'B',
        options: {
          title: String,
          color: String,
        },
      };
    }

    class C extends B {
      static config = {
        name: 'C',
        options: {
          color: Boolean,
        },
      };
    }

    class D extends C {
      static config = {
        name: 'D',
        log: false,
      };
    }

    const d = new D(h('div'));
    expect(d.__config).toMatchSnapshot();
  });

  it('should have a `$root` property', async () => {
    class ChildComponent extends Base {
      static config: BaseConfig = {
        name: 'ChildComponent',
      };
    }

    class Component extends Base<{ $children: { ChildComponent: ChildComponent[] } }> {
      static config: BaseConfig = {
        name: 'Component',
        components: { ChildComponent },
      };
    }

    class App extends Base<{ $children: { Component: Component[] } }> {
      static config: BaseConfig = {
        name: 'App',
        components: { Component },
      };
    }

    const div = h('div', {}, [
      h('div', { dataComponent: 'Component' }, [h('div', { dataComponent: 'ChildComponent' })]),
    ]);
    const app = new App(div).$mount();
    await advanceTimersByTimeAsync(1);
    expect(app.$root).toBe(app);
    expect(app.$children.Component[0].$root).toBe(app);
    expect(app.$children.Component[0].$children.ChildComponent[0].$root).toBe(app);
  });
});

describe('A Base instance methods', () => {
  it('should emit a mounted event', async () => {
    const { foo } = await getContext();
    const fn = mock();
    foo.$on('mounted', fn);
    foo.$destroy();
    await advanceTimersByTimeAsync(1);
    foo.$mount();
    await advanceTimersByTimeAsync(1);
    expect(fn).toHaveBeenCalledTimes(1);
    foo.$mount();
    await advanceTimersByTimeAsync(1);
    expect(fn).toHaveBeenCalledTimes(1);
  });

  it('should emit a destroyed event', async () => {
    const { foo } = await getContext();
    const fn = mock();
    foo.$on('destroyed', fn);
    foo.$destroy();
    await advanceTimersByTimeAsync(1);
    expect(fn).toHaveBeenCalledTimes(1);
    foo.$destroy();
    await advanceTimersByTimeAsync(1);
    expect(fn).toHaveBeenCalledTimes(1);
  });

  it('should be able to update its child components.', async () => {
    const { Foo } = await getContext();

    const div = h('div');
    div.innerHTML = `
      <div data-component="Bar"></div>
      <div data-component="Bar"></div>
    `;

    const fn = mock();

    class Bar extends Foo {}
    class Baz extends Base {
      static config: BaseConfig = { name: 'Baz' };

      updated() {
        fn(this.$id);
      }
    }

    class AsyncBaz extends Baz {
      static config: BaseConfig = {
        name: 'AsyncBaz',
      };
    }

    class App extends Base<{
      $children: {
        Bar: Bar[];
        Baz: Baz[];
        AsyncBaz: Promise<AsyncBaz>[];
      };
    }> {
      static config: BaseConfig = {
        name: 'App',
        components: {
          Bar,
          Baz,
          AsyncBaz: () => Promise.resolve(AsyncBaz),
        },
      };
    }

    const app = new App(div).$mount();
    await advanceTimersByTimeAsync(1);
    expect(app.$children.Bar).toHaveLength(2);
    expect(app.$children.Bar[0].$isMounted).toBe(true);
    div.innerHTML = `
      <div data-component="Baz"></div>
      <div data-component="Baz"></div>
      <div data-component="AsyncBaz"></div>
    `;

    app.$update();
    await advanceTimersByTimeAsync(1);

    expect(app.$children.Bar).toEqual([]);
    expect(app.$children.Baz).toHaveLength(2);
    expect(app.$children.Baz[0].$isMounted).toBe(true);
    const asyncBaz = await app.$children.AsyncBaz[0];
    expect(asyncBaz.$isMounted).toBe(true);

    const id = getInstanceFromElement(div.firstElementChild as HTMLElement, Baz).$id;
    expect(id).toBe(app.$children.Baz[0].$id);

    const child = document.createElement('div');
    child.dataset.component = 'Baz';
    div.append(child);
    expect(id).toBe(app.$children.Baz[0].$id);

    app.$update();
    await advanceTimersByTimeAsync(1);

    expect(id).toBe(app.$children.Baz[0].$id);
    expect(fn).toHaveBeenCalledTimes(3);
  });

  it('should implement the $register method', async () => {
    const { Foo } = await getContext();

    class Bar extends Foo {}
    class Baz extends Foo {}
    class Boz extends Foo {}
    document.body.innerHTML = `
      <div data-component="Foo"></div>
      <div data-component="Bar"></div>
      <div data-component="Bar"></div>
      <div class="custom-selector"></div>
      <div class="custom-selector"></div>
    `;

    const fooInstances = Bar.$register();
    const barInstances = Bar.$register('Bar');
    const bazInstances = Baz.$register('.custom-selector');
    const bozInstances = Boz.$register('Boz');
    await advanceTimersByTimeAsync(1);
    expect(fooInstances).toHaveLength(1);
    expect(fooInstances[0].$el.dataset.component).toBe('Foo');
    expect(barInstances).toHaveLength(2);
    expect(barInstances[0] instanceof Bar).toBe(true);
    expect(bazInstances).toHaveLength(2);
    expect(bazInstances[0] instanceof Baz).toBe(true);
    expect(bozInstances).toHaveLength(0);
  });

  it('should be able to be terminated', async () => {
    const fn = mock();

    class Bar extends Base {
      static config: BaseConfig = {
        name: 'Bar',
      };

      terminated() {
        fn('method');
      }
    }

    const div = h('div');
    const bar = new Bar(div);
    bar.$mount();
    await advanceTimersByTimeAsync(1);
    expect(bar).toEqual(getInstanceFromElement(div, Bar));
    bar.$on('terminated', () => fn('event'));
    bar.$terminate();
    await advanceTimersByTimeAsync(1);
    expect(fn).toHaveBeenCalledTimes(2);
    expect(fn).toHaveBeenNthCalledWith(1, 'event');
    expect(fn).toHaveBeenNthCalledWith(2, 'method');
  });

  it('should not find children if none provided', async () => {
    const { foo } = await getContext();

    class Bar extends Base {
      static config: BaseConfig = {
        name: 'Bar',
      };
    }

    class Baz extends Base<{ $children: { Bar: Bar[] } }> {
      static config: BaseConfig = {
        name: 'Baz',
        components: { Bar },
      };
    }

    expect(foo.$children.registeredNames).toEqual([]);
    const baz = new Baz(h('div'));
    baz.$mount();
    await advanceTimersByTimeAsync(1);
    expect(baz.$children.Bar).toEqual([]);
  });

  it('should not find terminated children', async () => {
    class Bar extends Base {
      static config = {
        name: 'Bar',
      };
    }

    class Baz extends Base {
      static config = {
        name: 'Baz',
        components: { Bar },
      };
    }

    const div = h('div');
    div.innerHTML = '<div data-component="Bar"></div>';

    const baz = new Baz(div);
    baz.$mount();
    await advanceTimersByTimeAsync(1);
    const bar = getInstanceFromElement(div.firstElementChild as HTMLElement, Bar);
    expect(baz.$children.Bar).toEqual([bar]);
    bar.$terminate();
    await advanceTimersByTimeAsync(1);
    expect(baz.$children.Bar).toEqual([]);
  });

  it('should mount and destroy its children', async () => {
    const { Foo } = await getContext();
    class Bar extends Base {
      static config: BaseConfig = {
        name: 'Bar',
      };
    }
    class Baz extends Foo<{ $children: { Bar: Bar[] } }> {
      static config: BaseConfig = {
        name: 'Baz',
        components: { Bar },
      };
    }

    document.body.innerHTML = '<div data-component="Bar"></div>';
    const baz = new Baz(document.body);
    baz.$mount();
    await advanceTimersByTimeAsync(1);
    const barElement = document.querySelector('[data-component="Bar"]') as HTMLElement;
    expect(baz.$isMounted).toBe(true);
    expect(getInstanceFromElement(barElement, Bar).$isMounted).toBe(true);
    baz.$destroy();
    await advanceTimersByTimeAsync(1);
    expect(baz.$isMounted).toBe(false);
    expect(getInstanceFromElement(barElement, Bar).$isMounted).toBe(false);
  });
});

describe('The Base class event methods', () => {
  it('should bind handlers to events', async () => {
    class App extends Base {
      static config: BaseConfig = {
        name: 'A',
        emits: ['foo'],
      };
    }

    const app = new App(document.createElement('div'));
    app.$mount();
    await advanceTimersByTimeAsync(1);
    const fn = mock();
    const off = app.$on('foo', fn);
    app.$emit('foo', { foo: true });
    expect(fn).toHaveBeenCalledTimes(1);
    expect(fn).toHaveBeenLastCalledWith(
      expect.objectContaining({ type: 'foo', detail: [{ foo: true }] }),
    );
    off();
    app.$emit('foo', { foo: true });
    expect(fn).toHaveBeenCalledTimes(1);
  });

  it('should store event handlers', async () => {
    class App extends Base {
      static config = { name: 'App', emits: ['event'] };
    }

    const app = new App(h('div'));
    app.$mount();
    await advanceTimersByTimeAsync(1);
    const fn = mock();
    app.$on('event', fn);
    expect(app.__hasEvent('event')).toBe(true);
    expect(app.__eventHandlers.get('event')).toEqual(new Set([fn]));
    app.$off('event', fn);
    expect(app.__eventHandlers.get('event')).toEqual(new Set());
  });

  it('should warn when an event is not configured', async () => {
    class App extends Base {
      static config = { name: 'App' };
    }

    const app = new App(document.createElement('div'));
    app.$mount();
    await advanceTimersByTimeAsync(1);
    const warnMock = spyOn(console, 'warn');
    warnMock.mockImplementation(() => {});
    app.$on('other-event', () => {});
    expect(warnMock).toHaveBeenCalledTimes(1);
    warnMock.mockRestore();
  });
});

describe('A Base instance config', () => {
  let element;

  beforeEach(() => {
    element = h('div');
  });

  it('should have a working $log method when active', () => {
    class Foo extends Base {
      static config = {
        name: 'Foo',
        log: true,
      };
    }
    const spy = spyOn(window.console, 'log');
    spy.mockImplementation(() => true);
    const foo = new Foo(element).$mount();
    expect(foo.$options.log).toBe(true);
    foo.$log('bar');
    expect(spy).toHaveBeenCalledWith('[Foo]', 'bar');
    spy.mockRestore();
  });

  it('should have a silent $log method when not active', () => {
    class Foo extends Base {
      static config = {
        name: 'Foo',
        log: false,
      };
    }
    const spy = spyOn(window.console, 'log');
    const foo = new Foo(element).$mount();
    expect(foo.$options.log).toBe(false);
    foo.$log('bar');
    expect(spy).toHaveBeenCalledTimes(0);
    spy.mockRestore();
  });

  it('should have a working $warn method when active', () => {
    class Foo extends Base {
      static config = {
        name: 'Foo',
        log: true,
      };
    }
    const spy = spyOn(window.console, 'warn');
    spy.mockImplementation(() => true);
    const foo = new Foo(element).$mount();
    expect(foo.$options.log).toBe(true);
    foo.$warn('bar');
    expect(spy).toHaveBeenCalledWith('[Foo]', 'bar');
    spy.mockRestore();
  });

  it('should have a silent $warn method when not active', () => {
    class Foo extends Base {
      static config = {
        name: 'Foo',
        log: false,
      };
    }
    const spy = spyOn(window.console, 'warn');
    const foo = new Foo(element).$mount();
    expect(foo.$options.log).toBe(false);
    foo.$warn('bar');
    expect(spy).toHaveBeenCalledTimes(0);
    spy.mockRestore();
  });

  it('should have a working debug method when active in dev mode', () => {
    class Foo extends Base {
      static config = {
        name: 'Foo',
        debug: true,
      };
    }
    globalThis.__DEV__ = true;
    process.env.NODE_ENV = 'development';
    const spy = spyOn(window.console, 'log');
    spy.mockImplementation(() => true);
    const div = document.createElement('div');
    new Foo(div).$mount();
    for (const args of spy.mock.calls) {
      expect(args[0]).toStartWith('[debug] [Foo');
    }
    spy.mockRestore();
    process.env.NODE_ENV = 'test';
  });
});
