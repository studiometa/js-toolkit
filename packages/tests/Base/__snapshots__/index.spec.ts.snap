// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`A Base instance should inherit from parent config 1`] = `
{
  "components": {},
  "emits": [
    "before-mounted",
    "mounted",
    "updated",
    "destroyed",
    "terminated",
    "ticked",
    "scrolled",
    "resized",
    "moved",
    "loaded",
    "keyed",
    "before-mounted",
    "mounted",
    "updated",
    "destroyed",
    "terminated",
    "ticked",
    "scrolled",
    "resized",
    "moved",
    "loaded",
    "keyed",
  ],
  "log": false,
  "name": "D",
  "options": {
    "color": [class Boolean],
    "title": [class String],
  },
  "refs": [],
}
`;

exports[`A Base instance config should have a working debug method when active in dev mode 1`] = `[]`;

exports[`A Base instance should inherit from parent config 2`] = `
{
  "components": {},
  "emits": [
    "before-mounted",
    "mounted",
    "updated",
    "destroyed",
    "terminated",
    "ticked",
    "scrolled",
    "resized",
    "moved",
    "loaded",
    "keyed",
    "before-mounted",
    "mounted",
    "updated",
    "destroyed",
    "terminated",
    "ticked",
    "scrolled",
    "resized",
    "moved",
    "loaded",
    "keyed",
  ],
  "log": false,
  "name": "D",
  "options": {
    "color": [class Boolean],
    "title": [class String],
  },
  "refs": [],
}
`;

exports[`A Base instance should inherit from parent config 3`] = `
{
  "components": {},
  "emits": [
    "before-mounted",
    "mounted",
    "updated",
    "destroyed",
    "terminated",
    "ticked",
    "scrolled",
    "resized",
    "moved",
    "loaded",
    "keyed",
    "before-mounted",
    "mounted",
    "updated",
    "destroyed",
    "terminated",
    "ticked",
    "scrolled",
    "resized",
    "moved",
    "loaded",
    "keyed",
  ],
  "log": false,
  "name": "D",
  "options": {
    "color": [class Boolean],
    "title": [class String],
  },
  "refs": [],
}
`;

exports[`A Base instance should inherit from parent config 4`] = `
{
  "components": {},
  "emits": [
    "before-mounted",
    "mounted",
    "updated",
    "destroyed",
    "terminated",
    "ticked",
    "scrolled",
    "resized",
    "moved",
    "loaded",
    "keyed",
    "before-mounted",
    "mounted",
    "updated",
    "destroyed",
    "terminated",
    "ticked",
    "scrolled",
    "resized",
    "moved",
    "loaded",
    "keyed",
  ],
  "log": false,
  "name": "D",
  "options": {
    "color": [class Boolean],
    "title": [class String],
  },
  "refs": [],
}
`;

exports[`A Base instance should inherit from parent config 5`] = `
{
  "components": {},
  "emits": [
    "before-mounted",
    "mounted",
    "updated",
    "destroyed",
    "terminated",
    "ticked",
    "scrolled",
    "resized",
    "moved",
    "loaded",
    "keyed",
    "before-mounted",
    "mounted",
    "updated",
    "destroyed",
    "terminated",
    "ticked",
    "scrolled",
    "resized",
    "moved",
    "loaded",
    "keyed",
  ],
  "log": false,
  "name": "D",
  "options": {
    "color": [class Boolean],
    "title": [class String],
  },
  "refs": [],
}
`;

exports[`A Base instance should inherit from parent config 6`] = `
{
  "components": {},
  "emits": [
    "before-mounted",
    "mounted",
    "updated",
    "destroyed",
    "terminated",
    "ticked",
    "scrolled",
    "resized",
    "moved",
    "loaded",
    "keyed",
    "before-mounted",
    "mounted",
    "updated",
    "destroyed",
    "terminated",
    "ticked",
    "scrolled",
    "resized",
    "moved",
    "loaded",
    "keyed",
  ],
  "log": false,
  "name": "D",
  "options": {
    "color": [class Boolean],
    "title": [class String],
  },
  "refs": [],
}
`;

exports[`A Base instance should inherit from parent config 7`] = `
{
  "components": {},
  "emits": [
    "before-mounted",
    "mounted",
    "updated",
    "destroyed",
    "terminated",
    "ticked",
    "scrolled",
    "resized",
    "moved",
    "loaded",
    "keyed",
    "before-mounted",
    "mounted",
    "updated",
    "destroyed",
    "terminated",
    "ticked",
    "scrolled",
    "resized",
    "moved",
    "loaded",
    "keyed",
  ],
  "log": false,
  "name": "D",
  "options": {
    "color": [class Boolean],
    "title": [class String],
  },
  "refs": [],
}
`;

exports[`A Base instance should inherit from parent config 8`] = `
{
  "components": {},
  "emits": [
    "before-mounted",
    "mounted",
    "updated",
    "destroyed",
    "terminated",
    "ticked",
    "scrolled",
    "resized",
    "moved",
    "loaded",
    "keyed",
    "before-mounted",
    "mounted",
    "updated",
    "destroyed",
    "terminated",
    "ticked",
    "scrolled",
    "resized",
    "moved",
    "loaded",
    "keyed",
  ],
  "log": false,
  "name": "D",
  "options": {
    "color": [class Boolean],
    "title": [class String],
  },
  "refs": [],
}
`;

exports[`A Base instance should inherit from parent config 9`] = `
{
  "components": {},
  "emits": [
    "before-mounted",
    "mounted",
    "updated",
    "destroyed",
    "terminated",
    "ticked",
    "scrolled",
    "resized",
    "moved",
    "loaded",
    "keyed",
    "before-mounted",
    "mounted",
    "updated",
    "destroyed",
    "terminated",
    "ticked",
    "scrolled",
    "resized",
    "moved",
    "loaded",
    "keyed",
  ],
  "log": false,
  "name": "D",
  "options": {
    "color": [class Boolean],
    "title": [class String],
  },
  "refs": [],
}
`;
